CMakeLists.txt

```diff
@@ -25,7 +25,16 @@ endif()
 set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin )
 set( CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib )
 
+# Define macro to use atomic implemnetation with GCC built-in.
+# Up until GCC 4.6.3, __sync_* shall be called. From 4.7.0, 
+# __atomic_* are available. We only implemented atomic
+# with __atomic_*.
+#
+# todo: check compiler is GCC AND GCC version is above 4.7.0.
```

Comment:
one thing at a time. this is a solid todo.

We require CMake min version 3.5. A couple of macros I tried, haven't found a good solution. 
```CMAKE_COMPILER_IS_GNUCC``` --> works
```CMAKE_CXX_COMPILER_VERSION``` --> doesn't work
```CMAKE_CXX_COMPILER_ID``` --> doesn't work.

Reply:
Try `CMAKE_C_COMPILER_VERSION` and `CMAKE_C_COMPILER_ID`

(`_CXX_` is for C++ compilers)

Reply:
Also consider moving this out of the root CMakeLists.txt.  Since atomic is in platform, it should go in the platform CMakeLists.txt.

Reply:
+1

Reply:
Thanks. Will update to below once we finalize where to put atomic.h.
```
if ( CMAKE_COMPILER_IS_GNUCC AND CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 4.7 )
    add_definitions( -DCOMPILER_OPTION_GCC_ATOMIC_BUILTIN=1 )
endif()
```